Handle single rpc error responses for batch requests:

fselmo opened the pull request:
### What was wrong?

- With some clients, when executing a batch request, the client will return a single error response for the entire batch. Previously, we were assuming that batch requests always received a list of responses.

Closes #3518

### How was it fixed?

- This commit updates the batch request logic to handle the above case.
- Add tests for batch requests with a single error response.

### Todo:

- [x] Clean up commit history
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

#### Cute Animal Picture

🐈

fselmo committed:
Handle single rpc error responses for batch requests:

- With some clients, when executing a batch request, the client will
  return a single error response for the entire batch. Previously,
  we were assuming that batch requests always received a list of
  responses. This commit updates the batch request logic to handle
  this case.

- Add tests for batch requests with a single error response.

reedsa commented on tests/core/providers/test_websocket_provider.py:
```diff
@@ -432,3 +433,24 @@ async def test_listener_task_breaks_out_of_handle_subscriptions_when_cancelled()
+                with pytest.raises(Web3RPCError, match="empty batch"):
+                    await batch.async_execute()
+
+            # assert that event though there was an error, we have reset the batching
```
```suggestion
            # assert that even though there was an error, we have reset the batching
```

reedsa commented on tests/core/providers/test_http_provider.py:
```diff
@@ -110,3 +115,20 @@ def test_get_request_headers(provider):
+        with pytest.raises(Web3RPCError, match="empty batch"):
+            batch.execute()
+
+    # assert that event though there was an error, we have reset the batching state
```
```suggestion
    # assert that even though there was an error, we have reset the batching state
```

reedsa commented on tests/core/providers/test_async_ipc_provider.py:
```diff
@@ -366,3 +367,24 @@ async def test_async_ipc_provider_write_messages_end_with_new_line_delimiter(
+            with pytest.raises(Web3RPCError, match="empty batch"):
+                await batch.async_execute()
+
+        # assert that event though there was an error, we have reset the batching state
```
```suggestion
        # assert that even though there was an error, we have reset the batching state
```

reedsa commented on tests/core/providers/test_async_http_provider.py:
```diff
@@ -107,3 +112,21 @@ def test_get_request_headers(provider):
+        with pytest.raises(Web3RPCError, match="empty batch"):
+            await batch.async_execute()
+
+    # assert that event though there was an error, we have reset the batching state
```
```suggestion
    # assert that even though there was an error, we have reset the batching state
```

reedsa commented on web3/manager.py:
```diff
@@ -436,15 +436,27 @@ def _make_batch_request(
+        if not isinstance(response, list):
+            # Expect a JSON-RPC error response and call _validate_response to raise
+            # the appropriate exception
+            _validate_response(
```
Is there a case that an exception is not raised here and would be validated again inside the `_format_batched_response` call below? If that happens, might be simpler to coerce the response to a list and do away with this extra step. I may not have context about the difference between passing empty params here vs. calling with params in the format method.

This change might also make the `responses` variable viable again.

fselmo replied:
I originally thought of coercing into a list but it seems that the JSON-RPC spec itself can return a single error object with null `id` for batch requests. In that sense, it seems it would be misleading to implement it as if it returned a list with a single object inside. In the case that a user decides to call `provider.make_batch_request()`, for example, they would get a more accurate response from the library rather than some other object that was not actually returned by the client.

That said, I think we could make a more proper check that it is indeed an `error` object inside the single response, otherwise we should probably invalidate the response since it's an unexpected result altogether. But maybe that check can live inside the `_validate_response` method.

fselmo replied:
I kept a more explicit check that will error out if some wrong formatting comes through. I think it's better to be explicit about out expectation here and fail hard if it isn't what we expect. Let me know your thoughts.

reedsa replied:
I agree on handling the error in the response explicitly, this looks good!

reedsa reviewed (commented):
One idea/question and a few trivial comment updates. 👍

fselmo committed:
Address comments on PR #3585:

- Type fixes
- Better organize error handling validation for batch error response

kclowes reviewed (approved):
LGTM! 🚀
